### ISSUE TITLE
wrapElement for issues like `Warning: validateDOMNesting(...): <div> cannot appear as a descendant of <p>.`

### ISSUE THREAD

**Issue by Songkeys** [rpearce/react-medium-image-zoom#356]
## Issue Type

- [ ] Bug report
- [x] Feature request
- [ ] Question / support request
- [ ] Other

## Description

May I know why `wrapElement`, which was added to solve problem like https://github.com/rpearce/react-medium-image-zoom/pull/238, was removed in v5?

Could we add it back or is there any other solution to this?

**Comment by rpearce** [comment]
@Songkeys Thanks for opening this issue!

When I rewrote this for v5, I forgot about this issue that using `<span>`s solved. There is more than 1 `<div>` now to consider. In order to resolve this, I have two paths that I can see at present (let me know if there are more):

1. Re-add an option to provide an element as a string before like `'span'`, but it'll need to apply to >1 element (might just be 2 elements, but still...)
2. Convert the `<div>`s to `<span>`, myself, and make sure they still work the same (ensuring `display` properties are good)

Let me know what you think after reading 👆 !

**Comment by Songkeys** [comment]
Option 1 LGTM.

Wouldn't the option 2 breaks the current behaviour anyway? Because someone might use `.xxx > div` to customise the style or manipulate the dom.

About the multiple `div`s issue, may i ask what the point of the double [wrappers](https://github.com/rpearce/react-medium-image-zoom/blob/213743fcb01e7a28ac7258f7be0857ffc4dbc685/source/Controlled.tsx#L181)? Are we able to remove the outer one?

**Comment by rpearce** [comment]
> About the multiple divs issue, may i ask what the point of the double [wrappers](https://github.com/rpearce/react-medium-image-zoom/blob/213743fcb01e7a28ac7258f7be0857ffc4dbc685/source/Controlled.tsx#L181) is? Are we able to remove the outer one?

Since the `children` can be practically anything, we need something to anchor on to, querySelector, and watch for DOM mutations: https://github.com/rpearce/react-medium-image-zoom/blob/213743fcb01e7a28ac7258f7be0857ffc4dbc685/source/Controlled.tsx#L275-L307

The outer `<div>` encapsulates and relatively positions the other elements, like the "ghost" (used for the keyboard button support).

> Wouldn't the option 2 break the current behaviour anyway? Because someone might use `.xxx > div` to customise the style or manipulate the dom.

I hope nobody is targeting like that when there are selectors like `[data-rmiz]` to select, but you're right that it could be happening.

* * *

I think you've helped make the answer clear, so the next question is this: what's the API? There are options:

1. Bring back `wrapElement` to make it an easier upgrade from `v4` to `v5`, but apply it to both elements
2. Use a boolean flag like `preferInline={true}` or something
3. ???

What do you think?

**Comment by Songkeys** [comment]
> 1. Bring back `wrapElement` to make it an easier upgrade from v4 to v5, but apply it to both elements

I prefer this one, which allows us to even pass something else like `"a"` or whatever. It adds convenience to the migration as well (but I'm not sure how much this would actually help though).

> 2. Use a boolean flag like `preferInline={true}` or something

This one looks semantic and problem-target. But it may be ambiguous with the meaning of "inline" - i.e., does it affect the element or only the styling (`display: inline`)?

**Comment by rpearce** [comment]
`wrapElement` it is! I'll get a PR up and in tonight, if I can, and I'll combine this with another feature that's been sitting for a while for a `5.1.0` release.

**Comment by rpearce** [comment]
@Songkeys If a `<p>` is wrapping around this image, it exposes another DOM nesting concern:

```
Warning: validateDOMNesting(...): <dialog> cannot appear as a descendant of <p>.
```

`v5` is using the `<dialog>` element, a block-level element, so the same issue arises if you're wrapping this in `<p></p>`.

It's 11:45pm at night, so I'm going to sleep on this.

**Comment by Songkeys** [comment]
> `<dialog>` cannot appear as a descendant of `<p>`.

Yes. This would happen if user passes `wrapElement="p"`. Oh I do wish there was a style-less and semantic-less element called `group`...

Do you think it's a good idea if we:

1. again, use `span` instead of `dialog`;
2. move `dialog` out and wrap them using a `Fragment`:

```jsx
<>
  <wrapElement> ... </wrapElement>
  {hasDialog && <dialog> ... </dialog>}
</>
```

**Comment by rpearce** [comment]
Good thoughts! Let's see...

`<dialog>` is required so that we don't run into all the accessibility issues of not having a `<dialog>`. There are _quite_ a few things that would need to be recreated, and I'd prefer not to have to add all that code when browsers now do it for us.

Even if it's outside of the `<span>`, it seems like you might be wrapping the entire thing in a `<p>`; is that accurate?

If so, my only thought would be try _try_ putting the `<dialog>` in a portal and placing it at the end of the `<body>`. It feels preposterous, but it's all I've got right now 🤔

**Comment by Songkeys** [comment]
> Even if it's outside of the `<span>`, it seems like you might be wrapping the entire thing in a <p>; is that accurate?

Oh, that's right. I totally forgot this again. Sorry.

Portal is a good idea.

Or, what do you think of "web-component", which should be able to be put inside `<p>` no matter what is inside its shadow root. The benefit over portal is that it's more intuitive. (I'm not sure which one requires more work though, which would turn out to be another dealbreaker.)

**Comment by rpearce** [comment]
Hilariously, the portal idea worked on the first try, and it works well with VoiceOver in Safari. I'll need to test the other screen readers, as well, but I can probably put up a beta release in the next day or two.

I've got https://github.com/rpearce/react-medium-image-zoom/pull/358 drafted in the mean time.

**Comment by rpearce** [comment]
FYI https://github.com/rpearce/react-medium-image-zoom/pull/358#issuecomment-1253148081

**Comment by rpearce** [comment]
@Songkeys This has been released in `v5.1.0`!

I added you to the All Contributors section in the README for your contributions. Thank you for your time 💜

**Comment by Songkeys** [comment]
Thank you! I just installed it and it works really well!
